File: server/src/conflictCenter.ts
// Three-way Conflict Center: index stages 1/2/3 inspection and resolution.
//
// Ported from `src-tauri/src/conflict_center.rs` (desktop module).

import fs from "node:fs/promises";
import path from "node:path";
import { AppError, invalidError } from "./lib/errors.ts";
import { checkedStdout, runGitRaw } from "./lib/git.ts";
import { repositoryOperationState } from "./operationState.ts";

const MAX_TEXT_BYTES = 2 * 1024 * 1024;

interface StageEntry {
  mode: string;
  oid: string;
}

interface ConflictEntries {
  base?: StageEntry;
  current?: StageEntry;
  incoming?: StageEntry;
}

export interface ConflictStage {
  oid: string;
  mode: string;
  kind: "blob" | "symlink" | "gitlink";
  text: string | null;
  binary: boolean;
  truncated: boolean;
  size: number;
}

export interface ConflictLabels {
  base: string;
  current: string;
  incoming: string;
}

export interface ConflictFileSummary {
  path: string;
  kind: ConflictKind;
  hasBase: boolean;
  hasCurrent: boolean;
  hasIncoming: boolean;
  binary: boolean;
  special: boolean;
  tooLarge: boolean;
}

export interface ConflictFileDetail {
  path: string;
  kind: ConflictKind;
  labels: ConflictLabels;
  base: ConflictStage | null;
  current: ConflictStage | null;
  incoming: ConflictStage | null;
  workingText: string | null;
  workingBinary: boolean;
  workingTruncated: boolean;
  workingSize: number;
}

export type ConflictKind =
  | "content"
  | "addAdd"
  | "currentDeleted"
  | "incomingDeleted"
  | "bothDeleted"
  | "complex";

export interface ConflictResolutionRequest {
  file: string;
  strategy: "current" | "incoming" | "merged" | "working" | "delete";
  content?: string | null;
}

export interface ConflictResolutionResult {
  remaining: number;
  nextPath: string | null;
}

type ConflictMap = Map<string, ConflictEntries>;

async function parseConflicts(inputPath: string): Promise<ConflictMap> {
  const output = await runGitRaw(inputPath, ["ls-files", "--unmerged", "-z", "--"]);
  if (output.code !== 0) {
    throw new AppError("command", `git ls-files --unmerged: ${output.stderr.trim()}`);
  }
  const conflicts: ConflictMap = new Map();
  for (const record of output.stdout.toString("utf8").split("\0")) {
    if (!record.length) continue;
    const text = record;
    const tab = text.indexOf("\t");
    if (tab < 0) continue;
    const header = text.slice(0, tab);
    const relative = text.slice(tab + 1);
    const fields = header.split(/\s+/).filter((value) => value.length > 0);
    if (fields.length < 3) continue;
    const [mode, oid] = fields;
    const stage = Number.parseInt(fields[2], 10);
    const entry: StageEntry = { mode, oid };
    let entries = conflicts.get(relative);
    if (!entries) {
      entries = {};
      conflicts.set(relative, entries);
    }
    if (stage === 1) entries.base = entry;
    else if (stage === 2) entries.current = entry;
    else if (stage === 3) entries.incoming = entry;
  }
  return conflicts;
}

// Mirrors Rust `path_is_safe`: no absolute path, no `.`/`..` components.
function pathIsSafe(relative: string): boolean {
  if (path.isAbsolute(relative)) return false;
  return relative.split(/[\\/]/).every((part) => part !== "." && part !== "..");
}

function stageKind(mode: string): ConflictStage["kind"] {
  if (mode === "160000") return "gitlink";
  if (mode === "120000") return "symlink";
  return "blob";
}

async function blobSize(inputPath: string, oid: string): Promise<number> {
  const output = await runGitRaw(inputPath, ["cat-file", "-s", oid]);
  if (output.code !== 0) {
    throw new (await import("./lib/errors")).AppError(
      "command",
      `git cat-file -s ${oid}: ${output.stderr.trim()}`,
    );
  }
  const size = Number.parseInt(output.stdout.toString("utf8").trim(), 10);
  if (!Number.isFinite(size)) {
    throw new AppError("invalid", `invalid blob size for ${oid}`);
  }
  return size;
}

function isBinaryBuffer(bytes: Buffer): boolean {
  if (bytes.includes(0)) return true;
  try {
    // Round-trip check equivalent to Rust `str::from_utf8`.
    new TextDecoder("utf-8", { fatal: true }).decode(bytes);
    return false;
  } catch {
    return true;
  }
}

async function stageDetail(inputPath: string, entry: StageEntry): Promise<ConflictStage> {
  const kind = stageKind(entry.mode);
  if (kind === "gitlink") {
    return { oid: entry.oid, mode: entry.mode, kind, text: null, binary: false, truncated: false, size: 0 };
  }
  const size = await blobSize(inputPath, entry.oid);
  if (size > MAX_TEXT_BYTES) {
    return { oid: entry.oid, mode: entry.mode, kind, text: null, binary: false, truncated: true, size };
  }
  const output = await runGitRaw(inputPath, ["cat-file", "blob", entry.oid]);
  if (output.code !== 0) {
    throw new AppError("command", `git cat-file blob ${entry.oid}: ${output.stderr.trim()}`);
  }
  const binary = isBinaryBuffer(output.stdout);
  const text = binary ? null : output.stdout.toString("utf8");
  return { oid: entry.oid, mode: entry.mode, kind, text, binary, truncated: false, size };
}

function conflictKind(entries: ConflictEntries): ConflictKind {
  const b = entries.base !== undefined;
  const c = entries.current !== undefined;
  const i = entries.incoming !== undefined;
  if (!b && c && i) return "addAdd";
  if (b && !c && i) return "currentDeleted";
  if (b && c && !i) return "incomingDeleted";
  if (b && c && i) return "content";
  if (b && !c && !i) return "bothDeleted";
  return "complex";
}

async function labels(inputPath: string): Promise<ConflictLabels> {
  const operation = (await repositoryOperationState(inputPath).catch(() => null))?.operation ?? null;
  const [current, incoming] =
    operation === "rebase"
      ? ["Current base", "Replayed commit"]
      : operation === "cherryPick"
        ? ["Current branch", "Cherry-picked commit"]
        : operation === "revert"
          ? ["Current branch", "Revert result"]
          : operation === "merge"
            ? ["Current branch", "Incoming branch"]
            : ["Current version", "Incoming version"];
  return { base: "Base", current, incoming };
}

// Mirrors Rust `safe_working_target`: canonicalize an existing ancestor and
// verify the target cannot escape the repository root through symlink parents.
async function safeWorkingTarget(root: string, relative: string): Promise<string> {
  const canonicalRoot = await fs.realpath(root);
  const target = path.join(root, relative);
  let ancestor = path.dirname(target);
  while (true) {
    try {
      await fs.lstat(ancestor);
      break;
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
      const parent = path.dirname(ancestor);
      if (parent === ancestor) {
        throw invalidError("conflict path has no existing repository ancestor");
      }
      ancestor = parent;
    }
  }
  const canonicalAncestor = await fs.realpath(ancestor);
  if (
    canonicalAncestor !== canonicalRoot &&
    !canonicalAncestor.startsWith(canonicalRoot + path.sep)
  ) {
    throw invalidError("conflict path escapes the repository through a symbolic-link parent");
  }
  return target;
}

async function readWorking(
  root: string,
  relative: string,
): Promise<{ text: string | null; binary: boolean; truncated: boolean; size: number }> {
  const file = await safeWorkingTarget(root, relative);
  let metadata: import("node:fs").Stats;
  try {
    metadata = await fs.lstat(file);
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      return { text: null, binary: false, truncated: false, size: 0 };
    }
    throw error;
  }
  if (metadata.isSymbolicLink()) return { text: null, binary: false, truncated: false, size: 0 };
  if (!metadata.isFile()) return { text: null, binary: true, truncated: false, size: metadata.size };
  const size = metadata.size;
  if (size > MAX_TEXT_BYTES) return { text: null, binary: false, truncated: true, size };
  const bytes = await fs.readFile(file);
  const binary = isBinaryBuffer(bytes);
  return { text: binary ? null : bytes.toString("utf8"), binary, truncated: false, size };
}

async function summarize(
  inputPath: string,
  relative: string,
  entries: ConflictEntries,
): Promise<ConflictFileSummary> {
  let special = false;
  let tooLarge = false;
  for (const entry of [entries.base, entries.current, entries.incoming]) {
    if (!entry) continue;
    const kind = stageKind(entry.mode);
    special = special || kind !== "blob";
    if (kind !== "gitlink") {
      tooLarge = tooLarge || (await blobSize(inputPath, entry.oid)) > MAX_TEXT_BYTES;
    }
  }
  return {
    path: relative,
    kind: conflictKind(entries),
    hasBase: entries.base !== undefined,
    hasCurrent: entries.current !== undefined,
    hasIncoming: entries.incoming !== undefined,
    binary: false,
    special,
    tooLarge,
  };
}

export async function repositoryConflicts(inputPath: string): Promise<ConflictFileSummary[]> {
  const conflicts = await parseConflicts(inputPath);
  const result: ConflictFileSummary[] = [];
  for (const [relative, entries] of conflicts) {
    result.push(await summarize(inputPath, relative, entries));
  }
  return result;
}

export async function conflictDetail(inputPath: string, relative: string): Promise<ConflictFileDetail> {
  if (!pathIsSafe(relative)) {
    throw invalidError("conflict path is not a safe repository-relative path");
  }
  const conflicts = await parseConflicts(inputPath);
  const entries = conflicts.get(relative);
  if (!entries) throw invalidError(`${relative} is not currently conflicted`);
  const root = await repositoryRootFor(inputPath);
  const working = await readWorking(root, relative);
  const [base, current, incoming] = await Promise.all([
    entries.base ? stageDetail(inputPath, entries.base) : Promise.resolve(null),
    entries.current ? stageDetail(inputPath, entries.current) : Promise.resolve(null),
    entries.incoming ? stageDetail(inputPath, entries.incoming) : Promise.resolve(null),
  ]);
  return {
    path: relative,
    kind: conflictKind(entries),
    labels: await labels(inputPath),
    base,
    current,
    incoming,
    workingText: working.text,
    workingBinary: working.binary,
    workingTruncated: working.truncated,
    workingSize: working.size,
  };
}

async function repositoryRootFor(inputPath: string): Promise<string> {
  return (await checkedStdout(inputPath, ["rev-parse", "--show-toplevel"])).trim();
}

function selectedStage(entries: ConflictEntries, strategy: string): StageEntry | null {
  if (strategy === "current") return entries.current ?? null;
  if (strategy === "incoming") return entries.incoming ?? null;
  return null;
}

async function removeWorkingPath(root: string, relative: string): Promise<void> {
  const target = await safeWorkingTarget(root, relative);
  let metadata: import("node:fs").Stats;
  try {
    metadata = await fs.lstat(target);
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") return;
    throw error;
  }
  if (metadata.isDirectory() && !metadata.isSymbolicLink()) {
    throw invalidError("refusing to recursively remove a directory while resolving a file conflict");
  }
  await fs.rm(target);
}

async function stageDeletion(
  inputPath: string,
  root: string,
  relative: string,
  gitlink: boolean,
): Promise<void> {
  if (gitlink) {
    await checkedStdout(inputPath, ["update-index", "--force-remove", "--", relative]);
    return;
  }
  await removeWorkingPath(root, relative);
  await checkedStdout(inputPath, ["add", "-A", "--", relative]);
}

async function resolveSide(
  inputPath: string,
  root: string,
  relative: string,
  entries: ConflictEntries,
  strategy: string,
): Promise<void> {
  const stage = selectedStage(entries, strategy);
  if (!stage) {
    const gitlink = [entries.base, entries.current, entries.incoming].some(
      (entry) => entry && stageKind(entry.mode) === "gitlink",
    );
    await stageDeletion(inputPath, root, relative, gitlink);
    return;
  }
  if (stageKind(stage.mode) === "gitlink") {
    await checkedStdout(inputPath, ["update-index", "--add", "--cacheinfo", stage.mode, stage.oid, relative]);
    return;
  }
  const stageNumber = strategy === "current" ? "2" : "3";
  await checkedStdout(inputPath, ["checkout-index", `--stage=${stageNumber}`, "--force", "--", relative]);
  await checkedStdout(inputPath, ["add", "-A", "--", relative]);
}

async function applyExecutableMode(target: string, entries: ConflictEntries): Promise<void> {
  if (process.platform === "win32") return;
  const executable = [entries.current, entries.incoming, entries.base].some(
    (entry) => entry && entry.mode === "100755",
  );
  if (!executable) return;
  try {
    const permissions = await fs.stat(target);
    if (!permissions.isFile()) return;
    await fs.chmod(target, 0o755);
  } catch {
    // best-effort, matching Rust's ignore-on-error behavior
  }
}

async function resolveMerged(
  inputPath: string,
  root: string,
  relative: string,
  entries: ConflictEntries,
  content: string,
): Promise<void> {
  const special = [entries.base, entries.current, entries.incoming].some(
    (entry) => entry && stageKind(entry.mode) !== "blob",
  );
  if (special) {
    throw invalidError(
      "manual text editing is disabled for symlink and submodule conflicts; choose a side instead",
    );
  }
  const target = await safeWorkingTarget(root, relative);
  await fs.mkdir(path.dirname(target), { recursive: true });
  try {
    const metadata = await fs.lstat(target);
    if (metadata.isSymbolicLink()) await fs.rm(target);
  } catch {
    // missing file: nothing to replace
  }
  await fs.writeFile(target, content);
  await applyExecutableMode(target, entries);
  await checkedStdout(inputPath, ["add", "-A", "--", relative]);
}

export async function resolveConflict(
  inputPath: string,
  request: ConflictResolutionRequest,
): Promise<ConflictResolutionResult> {
  if (!pathIsSafe(request.file)) {
    throw invalidError("conflict path is not a safe repository-relative path");
  }
  const conflicts = await parseConflicts(inputPath);
  const entries = conflicts.get(request.file);
  if (!entries) throw invalidError(`${request.file} is not currently conflicted`);
  const root = await repositoryRootFor(inputPath);

  switch (request.strategy) {
    case "current":
    case "incoming":
      await resolveSide(inputPath, root, request.file, entries, request.strategy);
      break;
    case "merged": {
      const content = request.content;
      if (content === undefined || content === null) {
        throw invalidError("merged resolution requires content");
      }
      await resolveMerged(inputPath, root, request.file, entries, content);
      break;
    }
    case "working":
      await checkedStdout(inputPath, ["add", "-A", "--", request.file]);
      break;
    case "delete": {
      const gitlink = [entries.base, entries.current, entries.incoming].some(
        (entry) => entry && stageKind(entry.mode) === "gitlink",
      );
      await stageDeletion(inputPath, root, request.file, gitlink);
      break;
    }
    default:
      throw invalidError(`unknown conflict resolution strategy: ${request.strategy}`);
  }

  const remaining = await parseConflicts(inputPath);
  const keys = [...remaining.keys()].sort();
  const nextPath =
    keys.find((candidate) => candidate > request.file) ?? keys[0] ?? null;
  return { remaining: keys.length, nextPath };
}

File: src-tauri/src/conflict_center.rs
use crate::error::{AppError, AppResult};
use crate::models::{
    ConflictFileDetail, ConflictFileSummary, ConflictLabels, ConflictResolutionRequest,
    ConflictResolutionResult, ConflictStage,
};

#[cfg(not(target_os = "android"))]
mod desktop {
    use std::collections::BTreeMap;
    use std::fs;
    use std::path::{Component, Path, PathBuf};
    use std::process::{Command, Output};

    use super::*;

    const MAX_TEXT_BYTES: usize = 2 * 1024 * 1024;

    #[derive(Debug, Clone)]
    struct StageEntry {
        mode: String,
        oid: String,
    }

    #[derive(Debug, Clone, Default)]
    struct ConflictEntries {
        base: Option<StageEntry>,
        current: Option<StageEntry>,
        incoming: Option<StageEntry>,
    }

    fn run_git(path: &str, args: &[&str]) -> AppResult<Output> {
        let mut command = Command::new("git");
        command.arg("-C").arg(path).args(args);
        command.env("GIT_TERMINAL_PROMPT", "0");
        Ok(command.output()?)
    }

    fn checked_git(path: &str, args: &[&str]) -> AppResult<Output> {
        let output = run_git(path, args)?;
        if output.status.success() {
            Ok(output)
        } else {
            Err(AppError::Command(format!(
                "git {}: {}",
                args.join(" "),
                String::from_utf8_lossy(&output.stderr).trim()
            )))
        }
    }

    fn repository_root(path: &str) -> AppResult<PathBuf> {
        let output = checked_git(path, &["rev-parse", "--show-toplevel"])?;
        Ok(PathBuf::from(String::from_utf8_lossy(&output.stdout).trim()))
    }

    fn parse_conflicts(path: &str) -> AppResult<BTreeMap<String, ConflictEntries>> {
        let output = checked_git(path, &["ls-files", "--unmerged", "-z", "--"])?;
        let mut conflicts: BTreeMap<String, ConflictEntries> = BTreeMap::new();

        for record in output.stdout.split(|byte| *byte == 0).filter(|record| !record.is_empty()) {
            let Some(tab) = record.iter().position(|byte| *byte == b'\t') else { continue; };
            let header = String::from_utf8_lossy(&record[..tab]);
            let path = String::from_utf8_lossy(&record[tab + 1..]).into_owned();
            let mut fields = header.split_whitespace();
            let Some(mode) = fields.next() else { continue; };
            let Some(oid) = fields.next() else { continue; };
            let Some(stage) = fields.next().and_then(|value| value.parse::<u8>().ok()) else { continue; };
            let entry = StageEntry { mode: mode.to_string(), oid: oid.to_string() };
            let stages = conflicts.entry(path).or_default();
            match stage {
                1 => stages.base = Some(entry),
                2 => stages.current = Some(entry),
                3 => stages.incoming = Some(entry),
                _ => {}
            }
        }

        Ok(conflicts)
    }

    fn path_is_safe(path: &str) -> bool {
        let candidate = Path::new(path);
        !candidate.is_absolute()
            && candidate.components().all(|component| matches!(component, Component::Normal(_)))
    }

    fn stage_kind(mode: &str) -> &'static str {
        match mode {
            "160000" => "gitlink",
            "120000" => "symlink",
            _ => "blob",
        }
    }

    fn blob_size(path: &str, oid: &str) -> AppResult<usize> {
        let output = checked_git(path, &["cat-file", "-s", oid])?;
        String::from_utf8_lossy(&output.stdout)
            .trim()
            .parse::<usize>()
            .map_err(|_| AppError::Invalid(format!("invalid blob size for {oid}")))
    }

    fn stage_detail(path: &str, entry: &StageEntry) -> AppResult<ConflictStage> {
        let kind = stage_kind(&entry.mode).to_string();
        if kind == "gitlink" {
            return Ok(ConflictStage {
                oid: entry.oid.clone(),
                mode: entry.mode.clone(),
                kind,
                text: None,
                binary: false,
                truncated: false,
                size: 0,
            });
        }

        let size = blob_size(path, &entry.oid)?;
        if size > MAX_TEXT_BYTES {
            return Ok(ConflictStage {
                oid: entry.oid.clone(),
                mode: entry.mode.clone(),
                kind,
                text: None,
                binary: false,
                truncated: true,
                size,
            });
        }

        let output = checked_git(path, &["cat-file", "blob", &entry.oid])?;
        let binary = output.stdout.iter().any(|byte| *byte == 0) || std::str::from_utf8(&output.stdout).is_err();
        let text = if binary { None } else { Some(String::from_utf8(output.stdout).unwrap_or_default()) };
        Ok(ConflictStage {
            oid: entry.oid.clone(),
            mode: entry.mode.clone(),
            kind,
            text,
            binary,
            truncated: false,
            size,
        })
    }

    fn conflict_kind(entries: &ConflictEntries) -> String {
        match (entries.base.is_some(), entries.current.is_some(), entries.incoming.is_some()) {
            (false, true, true) => "addAdd",
            (true, false, true) => "currentDeleted",
            (true, true, false) => "incomingDeleted",
            (true, true, true) => "content",
            (true, false, false) => "bothDeleted",
            _ => "complex",
        }.to_string()
    }

    fn labels(path: &str) -> ConflictLabels {
        let operation = crate::operation_state::repository_operation_state(path).ok().and_then(|state| state.operation);
        let (current, incoming) = match operation.as_deref() {
            Some("rebase") => ("Current base", "Replayed commit"),
            Some("cherryPick") => ("Current branch", "Cherry-picked commit"),
            Some("revert") => ("Current branch", "Revert result"),
            Some("merge") => ("Current branch", "Incoming branch"),
            _ => ("Current version", "Incoming version"),
        };
        ConflictLabels { base: "Base".into(), current: current.into(), incoming: incoming.into() }
    }

    fn safe_working_target(root: &Path, relative: &str) -> AppResult<PathBuf> {
        let canonical_root = fs::canonicalize(root)?;
        let target = root.join(relative);
        let mut ancestor = target.parent().unwrap_or(root);
        while !ancestor.exists() {
            ancestor = ancestor.parent().ok_or_else(|| AppError::Invalid("conflict path has no existing repository ancestor".into()))?;
        }
        let canonical_ancestor = fs::canonicalize(ancestor)?;
        if !canonical_ancestor.starts_with(&canonical_root) {
            return Err(AppError::Invalid("conflict path escapes the repository through a symbolic-link parent".into()));
        }
        Ok(target)
    }

    fn read_working(root: &Path, relative: &str) -> AppResult<(Option<String>, bool, bool, usize)> {
        let file = safe_working_target(root, relative)?;
        let metadata = match fs::symlink_metadata(&file) {
            Ok(metadata) => metadata,
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok((None, false, false, 0)),
            Err(error) => return Err(error.into()),
        };
        if metadata.file_type().is_symlink() {
            return Ok((None, false, false, 0));
        }
        if !metadata.is_file() {
            return Ok((None, true, false, metadata.len() as usize));
        }
        let size = metadata.len() as usize;
        if size > MAX_TEXT_BYTES {
            return Ok((None, false, true, size));
        }
        let bytes = fs::read(&file)?;
        let binary = bytes.iter().any(|byte| *byte == 0) || std::str::from_utf8(&bytes).is_err();
        let text = if binary { None } else { String::from_utf8(bytes).ok() };
        Ok((text, binary, false, size))
    }

    fn summarize(path: &str, relative: &str, entries: &ConflictEntries) -> AppResult<ConflictFileSummary> {
        let mut special = false;
        let mut too_large = false;
        for entry in [&entries.base, &entries.current, &entries.incoming].into_iter().flatten() {
            let kind = stage_kind(&entry.mode);
            special |= kind != "blob";
            if kind != "gitlink" {
                too_large |= blob_size(path, &entry.oid)? > MAX_TEXT_BYTES;
            }
        }
        Ok(ConflictFileSummary {
            path: relative.to_string(),
            kind: conflict_kind(entries),
            has_base: entries.base.is_some(),
            has_current: entries.current.is_some(),
            has_incoming: entries.incoming.is_some(),
            binary: false,
            special,
            too_large,
        })
    }

    pub fn repository_conflicts(path: &str) -> AppResult<Vec<ConflictFileSummary>> {
        parse_conflicts(path)?
            .iter()
            .map(|(relative, entries)| summarize(path, relative, entries))
            .collect()
    }

    pub fn conflict_detail(path: &str, relative: &str) -> AppResult<ConflictFileDetail> {
        if !path_is_safe(relative) {
            return Err(AppError::Invalid("conflict path is not a safe repository-relative path".into()));
        }
        let conflicts = parse_conflicts(path)?;
        let entries = conflicts.get(relative).ok_or_else(|| AppError::Invalid(format!("{relative} is not currently conflicted")))?;
        let root = repository_root(path)?;
        let (working_text, working_binary, working_truncated, working_size) = read_working(&root, relative)?;
        Ok(ConflictFileDetail {
            path: relative.to_string(),
            kind: conflict_kind(entries),
            labels: labels(path),
            base: entries.base.as_ref().map(|entry| stage_detail(path, entry)).transpose()?,
            current: entries.current.as_ref().map(|entry| stage_detail(path, entry)).transpose()?,
            incoming: entries.incoming.as_ref().map(|entry| stage_detail(path, entry)).transpose()?,
            working_text,
            working_binary,
            working_truncated,
            working_size,
        })
    }

    fn selected_stage<'a>(entries: &'a ConflictEntries, strategy: &str) -> Option<&'a StageEntry> {
        match strategy {
            "current" => entries.current.as_ref(),
            "incoming" => entries.incoming.as_ref(),
            _ => None,
        }
    }

    fn remove_working_path(root: &Path, relative: &str) -> AppResult<()> {
        let target = safe_working_target(root, relative)?;
        match fs::symlink_metadata(&target) {
            Ok(metadata) if metadata.is_dir() && !metadata.file_type().is_symlink() => {
                return Err(AppError::Invalid("refusing to recursively remove a directory while resolving a file conflict".into()));
            }
            Ok(_) => fs::remove_file(&target)?,
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {}
            Err(error) => return Err(error.into()),
        }
        Ok(())
    }

    fn stage_deletion(path: &str, root: &Path, relative: &str, gitlink: bool) -> AppResult<()> {
        if gitlink {
            checked_git(path, &["update-index", "--force-remove", "--", relative])?;
            return Ok(());
        }
        remove_working_path(root, relative)?;
        checked_git(path, &["add", "-A", "--", relative])?;
        Ok(())
    }

    fn resolve_side(path: &str, root: &Path, relative: &str, entries: &ConflictEntries, strategy: &str) -> AppResult<()> {
        let Some(stage) = selected_stage(entries, strategy) else {
            let gitlink = [entries.base.as_ref(), entries.current.as_ref(), entries.incoming.as_ref()]
                .into_iter().flatten().any(|entry| stage_kind(&entry.mode) == "gitlink");
            return stage_deletion(path, root, relative, gitlink);
        };

        if stage_kind(&stage.mode) == "gitlink" {
            checked_git(path, &["update-index", "--add", "--cacheinfo", &stage.mode, &stage.oid, relative])?;
            return Ok(());
        }

        let stage_number = if strategy == "current" { "2" } else { "3" };
        let stage_arg = format!("--stage={stage_number}");
        checked_git(path, &["checkout-index", &stage_arg, "--force", "--", relative])?;
        checked_git(path, &["add", "-A", "--", relative])?;
        Ok(())
    }

    fn apply_executable_mode(target: &Path, entries: &ConflictEntries) -> AppResult<()> {
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let executable = [entries.current.as_ref(), entries.incoming.as_ref(), entries.base.as_ref()]
                .into_iter().flatten().find(|entry| entry.mode == "100755").is_some();
            if executable && target.exists() {
                let mut permissions = fs::metadata(target)?.permissions();
                permissions.set_mode(0o755);
                fs::set_permissions(target, permissions)?;
            }
        }
        Ok(())
    }

    fn resolve_merged(path: &str, root: &Path, relative: &str, entries: &ConflictEntries, content: &str) -> AppResult<()> {
        let special = [entries.base.as_ref(), entries.current.as_ref(), entries.incoming.as_ref()]
            .into_iter().flatten().any(|entry| stage_kind(&entry.mode) != "blob");
        if special {
            return Err(AppError::Invalid("manual text editing is disabled for symlink and submodule conflicts; choose a side instead".into()));
        }
        let target = safe_working_target(root, relative)?;
        if let Some(parent) = target.parent() {
            fs::create_dir_all(parent)?;
        }
        if fs::symlink_metadata(&target).map(|metadata| metadata.file_type().is_symlink()).unwrap_or(false) {
            fs::remove_file(&target)?;
        }
        fs::write(&target, content.as_bytes())?;
        apply_executable_mode(&target, entries)?;
        checked_git(path, &["add", "-A", "--", relative])?;
        Ok(())
    }

    pub fn resolve_conflict(path: &str, request: &ConflictResolutionRequest) -> AppResult<ConflictResolutionResult> {
        if !path_is_safe(&request.file) {
            return Err(AppError::Invalid("conflict path is not a safe repository-relative path".into()));
        }
        let conflicts = parse_conflicts(path)?;
        let entries = conflicts.get(&request.file).ok_or_else(|| AppError::Invalid(format!("{} is not currently conflicted", request.file)))?;
        let root = repository_root(path)?;

        match request.strategy.as_str() {
            "current" | "incoming" => resolve_side(path, &root, &request.file, entries, &request.strategy)?,
            "merged" => {
                let content = request.content.as_deref().ok_or_else(|| AppError::Invalid("merged resolution requires content".into()))?;
                resolve_merged(path, &root, &request.file, entries, content)?;
            }
            "working" => { checked_git(path, &["add", "-A", "--", &request.file])?; }
            "delete" => {
                let gitlink = [entries.base.as_ref(), entries.current.as_ref(), entries.incoming.as_ref()]
                    .into_iter().flatten().any(|entry| stage_kind(&entry.mode) == "gitlink");
                stage_deletion(path, &root, &request.file, gitlink)?;
            }
            _ => return Err(AppError::Invalid(format!("unknown conflict resolution strategy: {}", request.strategy))),
        }

        let remaining = parse_conflicts(path)?;
        let next_path = remaining.keys()
            .find(|candidate| candidate.as_str() > request.file.as_str())
            .cloned()
            .or_else(|| remaining.keys().next().cloned());
        Ok(ConflictResolutionResult {
            remaining: remaining.len(),
            next_path,
        })
    }
}

#[cfg(target_os = "android")]
mod android {
    use super::*;

    pub fn repository_conflicts(_path: &str) -> AppResult<Vec<ConflictFileSummary>> {
        Err(AppError::Unsupported("the three-way Conflict Center is currently desktop-only".into()))
    }

    pub fn conflict_detail(_path: &str, _relative: &str) -> AppResult<ConflictFileDetail> {
        Err(AppError::Unsupported("the three-way Conflict Center is currently desktop-only".into()))
    }

    pub fn resolve_conflict(_path: &str, _request: &ConflictResolutionRequest) -> AppResult<ConflictResolutionResult> {
        Err(AppError::Unsupported("conflict resolution actions are currently desktop-only".into()))
    }
}

#[cfg(not(target_os = "android"))]
pub use desktop::{conflict_detail, repository_conflicts, resolve_conflict};
#[cfg(target_os = "android")]
pub use android::{conflict_detail, repository_conflicts, resolve_conflict};
